Docs: Update git workflow documentation (#25164)
WordPress/gutenberg: docs/contributors/git-workflow.md

# Git Workflow

The Gutenberg project follows a standard pull request process for contributions. See GitHub's documentation for [additional details about pull requests](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests).

> This documentation is intended to help you get started using git with Gutenberg. Git is a powerful source code management tool, to learn git deeply, check out the [Pro Git book](https://git-scm.com/book/en/v2) available free online under CC BY-NC-SA 3.0 license.

Reviewer: ```suggestion
This documentation is intended to help you get started using git with Gutenberg. Git is a powerful source code management tool; to learn git deeply, check out the [Pro Git book](https://git-scm.com/book/en/v2) available free online under CC BY-NC-SA 3.0 license.
```

I think it should grammatically be a semicolon there

An overview of the process for contributors is:

- Fork the Gutenberg repository.
- Clone the forked repository.
- Create a new branch.
- Make code changes.
- Confirm tests pass.
- Commit the code changes within the newly created branch.
- Push the branch to the forked repository.
- Submit a pull request to the Gutenberg repository.

See the [repository management document](/docs/contributors/repository-management.md) for additional information on how the Gutenberg project uses GitHub.

## Git Workflow Walkthrough

The workflow for code and documentation is the same—since both are managed in GitHub—so you can watch a [video walk-through of contributing documentation](https://wordpress.tv/2020/09/02/marcus-kazmierczak-contribute-developer-documentation-to-gutenberg/) and the accompanying [slides for contributing to Gutenberg](https://mkaz.blog/wordpress/contribute-documentation-to-gutenberg/).

Here is a visual overview of the Git workflow:

![Visual Overview of Git Workflow](https://developer.wordpress.org/files/2020/09/git-workflow.png)

**Step 1**: Go to the Gutenberg repository on GitHub and click Fork. This creates a copy of the main Gutenberg repository to your account.

> 

Reviewer 2: Should we add in what to do if a user has already forked gutenberg?

For instance I have perhaps 6-8 months ago forked gutenberg. I assume I need to update the fork? Or perhaps I should just delete my fork and fork it again? What would be the most natural way?

Author: :+1:

![Screenshot showing fork button on GitHub](https://developer.wordpress.org/files/2020/09/gutenberg-fork.png)

> **Step 2**: Clone your forked repository locally, it is located at: `https://github.com/YOUR-USER-NAME/gutenberg`. Cloning copies all the files to your computer. Open a terminal and run:

Reviewer: ```suggestion
**Step 2**: Clone your forked repository locally. It is located at: `https://github.com/YOUR-USER-NAME/gutenberg`. Cloning copies all the files to your computer. Open a terminal and run:
```

> 

Reviewer 2: I quit terminal and reopened it to be sure that I was in the root folder.

Reviewer 2: It seems the difference between focusing on only running a PR is this:
`git clone https://github.com/WordPress/gutenberg.git`

Compared to pushing. Creating a PR.
`git clone https://github.com/paaljoachim/gutenberg`

It looks like I will need to change the running a PR doc issue I am creating. https://github.com/WordPress/gutenberg/issues/25202
As it seems safe to first fork gutenberg repo and then clone that. Using a direct clone of the WP repo for pushing is not the correct way to go.

I will wait until I hear back from you and/or Noah.

```bash
git clone https://github.com/YOUR-USER-NAME/gutenberg
```

> This will create a directory called `gutenberg` with all the files for the project. It might take a couple of minutes, it is downloading the entire history of the Gutenberg project.

Reviewer: ```suggestion
This will create a directory called `gutenberg` with all the files for the project. It might take a couple of minutes because it is downloading the entire history of the Gutenberg project.
```

> **Step 3**: Create a branch for your change, see below for branch naming. For this example the branch name is the complete string: `update/my-branch`

Reviewer: ```suggestion
**Step 3**: Create a branch for your change (see below for branch naming). For this example, the branch name is the complete string: `update/my-branch`
```

```bash
git switch -c update/my-branch
```

**Step 4**: Make the code changes. Build, confirm, and test your change thoroughly. See [coding guidelines](/docs/contributors/coding-guidelines.md) and [testing overview](/docs/contributors/testing-overview.md) for guidance.

**Step 5**: Commit your change with a [good commmit message](https://make.wordpress.org/core/handbook/best-practices/commit-messages/). This will commit your change to your local copy of the repository.

```bash
git commit -m "Your Good Commit Message" path/to/FILE
```

> **Step 6**: Push your change up to GitHub. The change will be pushed to your copy of the repository on the GitHub

Reviewer: ```suggestion
**Step 6**: Push your change up to GitHub. The change will be pushed to your fork of the repository on the GitHub
```

probably worthwhile to use the same name for it here

```bash
git push -u origin upgrade/my-branch
```

> **Step 7**: Go to your repository on GitHub, it will automatically detect the change and give you a link to create a pull request.

Reviewer: ```suggestion
**Step 7**: Go to your forked repository on GitHub -- it will automatically detect the change and give you a link to create a pull request.
```

![Screenshot showing pull request link](https://developer.wordpress.org/files/2020/09/pull-request-create.png)

> **Step 8**: Create the pull request, this will create the request on the WordPress Gutenberg repository to integrate the change from your repository.

Reviewer: ```suggestion
**Step 8**: Create the pull request. This will create the request on the WordPress Gutenberg repository to integrate the change from your forked repository.
```

> **Step 9**: Keep up on the pull request, if any additional changes or updates are requested than make the changes locally and push them up, following Steps 4-6.

Reviewer: ```suggestion
**Step 9**: Keep up with new activity on the pull request. If any additional changes or updates are requested, then make the changes locally and push them up, following Steps 4-6.
```

Do not make a new pull request for updates; by pushing your change to your respotiroy it will update the same PR. In this sense, the PR is a pointer on the WordPress Gutenberg repository to your copy. So when you update your copy, the PR is also updated.

> That’s it, once approved and merged your copy will be incorporated into the main repository. 🎉

Reviewer: ```suggestion
That’s it! Once approved and merged, your change will be incorporated into the main repository. 🎉
```
